src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -0,0 +1,4787 @@
+		*is_target_path = 0;
+		return 0;
+	}
+
+	if (strncmp(szInput, ".", 2) == 0) {
+		pt_end = stpncpy(full_path_loc, cur_dir, DFS_MAX_PATH);
+		if ((long int)(pt_end - full_path_loc) >= DFS_MAX_PATH) {
+			printf("full_path_loc[] is not large enough.\nQuit\n");
+			exit(1);
+		}
+	} else if (szInput[0] == '/') {
+		pt_end = stpncpy(full_path_loc, szInput, DFS_MAX_PATH);
+		if ((long int)(pt_end - full_path_loc) >= DFS_MAX_PATH) {
+			printf("full_path_loc[] is not large enough.\nQuit\n");
+			exit(1);
+		}
+	} else {
+		/* relative path */
+		len = snprintf(full_path_loc, DFS_MAX_PATH, "%s/%s", cur_dir, szInput);
+		if (len >= DFS_MAX_PATH) {
+			printf("The length of path is too long.\n");
+			exit(1);
+		}
+	}
+
+	/* Assume full_path_loc[] = "/A/B/C/../D/E", it will be "/A/B/D/E" after remove_dot_dot. */
+	remove_dot_dot(full_path_loc);
+
+	/* Remove '/./'; Replace '//' with '/'; Remove '/' at the end of path. */
+	len = remove_dot_and_cleanup(full_path_loc);
+
+	idx_dfs = query_dfs_mount(full_path_loc);
+
+	if (idx_dfs >= 0) {
+		if (!daos_inited) {
+			rc = pthread_mutex_lock(&lock_init);
+			D_ASSERT(rc == 0);
+			rc = daos_init();
+			if (rc) {
+				printf("daos_init failed, rc: "DF_RC"\nQuit\n", DP_RC(rc));
+				exit(1);
+			}
+			daos_inited = true;
+			rc = pthread_mutex_unlock(&lock_init);
+			D_ASSERT(rc == 0);
+		}
+
+		*dfs_mt = &dfs_list[idx_dfs];
+		if ((*dfs_mt)->inited == 0) {
+			/* Need mutex before dfs init!!!!! */
+			if ((*dfs_mt)->pool == NULL) {
+				/* retrieve dfs, pool and container handles from dfuse */
+				rc = retrieve_handles_from_fuse(idx_dfs);
+				if (rc) {
+					/* Let the original functions to handle it. */
+					*is_target_path = 0;
+					return 0;
+				}
+			} else {
+				init_dfs(idx_dfs);
+			}
+			(*dfs_mt)->inited = 1;
+		}
+		*is_target_path = 1;
+
+		if (full_path)
+			strcpy(full_path, full_path_loc + (*dfs_mt)->len_fs_root);
+
+		/* root dir */
+		if (full_path_loc[(*dfs_mt)->len_fs_root] == 0) {
+			*parent       = NULL;
+			parent_dir[0] = '\0';
+			item_name[0]  = '/';
+			item_name[1]  = '\0';
+			if (full_path)
+				strcpy(full_path, "/");
+		} else {
+			for (pos = len - 1; pos >= (*dfs_mt)->len_fs_root; pos--) {
+				if (full_path_loc[pos] == '/')
+					break;
+			}
+			strcpy(item_name, full_path_loc + pos + 1);
+			/* the item under root directory */
+			if (pos == (*dfs_mt)->len_fs_root) {
+				*parent       = NULL;
+				parent_dir[0] = '/';
+				parent_dir[1] = '\0';
+			} else {
+				/* Need to look up the parent directory */
+				full_path_loc[pos] = 0;
+				strcpy(parent_dir, full_path_loc + (*dfs_mt)->len_fs_root);
+			}
+			*parent = lookup_insert_dir(idx_dfs, parent_dir, &mode);
+			if (*parent == NULL) {
+				/* parent dir does not exist or something wrong */
+				errno = ENOENT;
+				return (-1);
+			}
+		}
+	} else {
+		*is_target_path = 0;
+		*parent         = NULL;
+		item_name[0]    = '\0';
+	}
+	return 0;
+}
+
+static void
+remove_dot_dot(char path[])
```

Comment:
How does this handle filenames with .. in, for example "myhouse..jpg" vs "../myhouse.jpg"?  I'm struggling to see what would happen if the path traversal goes above the root here, for example "a/../../../"